src/support/small_set.h

```diff
@@ -132,16 +135,20 @@ class SmallSetBase {
   // fixed-space storage
   FixedStorage fixed;
 
-  // flexible additional storage
-  FlexibleSet flexible;
+  // Flexible additional storage. We use std::optional here to reduce the
+  // overhead in the common case, where the flexible storage is never needed.
+  // With std::optional we just have a pointer taking up space, as opposed to a
```

Comment:
Rereading this now, this doesn't seem true at least on linux, as `sizeof` tells me that the full storage is used. That is, the gnu C++ STL seems to implement an optional as enough space for the item, and a tag whether it is present or not.

Given that, I think we might have been wrong in our previous discussion here. That is, using `std::optional` here increases memory size, so it may have downsides, and we should probably be more careful.

If our goal is to avoid a problem with the MSVC STL, then a workaround specific to there might make more sense actually.

What I'd suggest immediately though is to perhaps go look at what LLVM's SmallSet does, as a lot of thought has gone into that. We should probably do what they are doing.

Reply:
Unfortunately, [LLVM's SmallSet](https://llvm.org/doxygen/SmallSet_8h_source.html) doesn't seem to address the heap allocation on MSVC. 

I believe the [measurements you had made](https://github.com/WebAssembly/binaryen/pull/4885#issuecomment-1209693350) were on Linux? It seemed benefits from using avoiding the std::set construction outweighed the size increase there, too? (At least when used in EffectsAnalyzer, and way less drastic as on windows... I suppose it's not making heap allocations in the STL implementation there).

Effectively, we could, of course, avoid the entire PR, if we'd use another STL, [as you had suggested in this comment](https://github.com/WebAssembly/binaryen/issues/4165#issuecomment-1176486775).